mantidproject/mantid
# Speed up PredictPeaks for non-rectangular detectors

~~**Must not be merged before #18936**~~

This speeds up `PredictPeaks` for non-rectangular detectors by using a KDTree of detector positions rather than doing a full ray tracing search over all instrument components.

**To test:**

There are three cases that need testing:
 - Instruments with Rectangular detectors (e.g. TOPAZ)
 - Instruments with non-rectangular detectors (e.g. WISH)
 - Instruments with non-rectangular detectors and the tube-gap parameter (currently just CORELLI)

Here's a python script that will run PredictPeaks on all three cases. Note that the time taken to run `PredictPeaks` in each case should remain the same for TOPAZ and decrease for WISH and CORELLI with these changes applied.

**Note:** CORELLI took >1hr to run prior to these changes, so maybe get a cup of coffee while that one is running.

```python
ws = CreateSimulationWorkspace(Instrument='CORELLI',
                               BinParams='0,10000,20000',
                               UnitX='TOF')              
SetUB(ws, a=5.5, b=6.5, c=8.1, u='12,1,1', v='0,4,9')
cs = CrystalStructure('5.5 6.5 8.1', 'P b c a',
                      """Ni 0.232 0.114 0.543 1.0 0.00843;
                         Al 0.434 0.041 0.854 1.0 0.01120""")
ws.sample().setCrystalStructure(cs)
predicted = PredictPeaks(InputWorkspace=ws,
                         CalculateStructureFactors=False,
                         MinDSpacing=0.1,
                         MaxDSpacing=15,
                         PredictPeaksOutsideDetectors=False,
                         WavelengthMin=0.5, WavelengthMax=10.2)

ws2 = CreateSimulationWorkspace(Instrument='TOPAZ',
                               BinParams='0,10000,20000',
                               UnitX='TOF')
SetUB(ws2, a=5.5, b=6.5, c=8.1, u='12,1,1', v='0,4,9')
ws2.sample().setCrystalStructure(cs)                    
predicted2 = PredictPeaks(InputWorkspace=ws2,
                         CalculateStructureFactors=False,
                         MinDSpacing=0.1,
                         MaxDSpacing=15,
                         PredictPeaksOutsideDetectors=False,
                         WavelengthMin=0.5, WavelengthMax=10.2)
                         
ws2 = CreateSimulationWorkspace(Instrument='WISH',
                               BinParams='0,10000,20000',
                               UnitX='TOF')
SetUB(ws2, a=5.5, b=6.5, c=8.1, u='12,1,1', v='0,4,9')
ws2.sample().setCrystalStructure(cs)                    
predicted2 = PredictPeaks(InputWorkspace=ws2,
                         CalculateStructureFactors=False,
                         MinDSpacing=0.1,
                         MaxDSpacing=15,
                         PredictPeaksOutsideDetectors=False,
                         WavelengthMin=0.5, WavelengthMax=10.2)
```

Fixes #19131 .

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
